lib/Target/LLVM/LLVMToDynamicLib.cpp

```diff
@@ -0,0 +1,313 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Target/LLVM/LLVMToDynamicLib.h"
+
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Target/LLVMIR/ModuleTranslation.h"
+
+// #include "lld/Common/Driver.h"
+// #include "lld/Common/ErrorHandler.h"
+// #include "lld/Common/Memory.h"
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Analysis/TargetLibraryInfo.h"
+#include "llvm/ExecutionEngine/ExecutionEngine.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Target/TargetMachine.h"
+#include "llvm/Target/TargetOptions.h"
+#include "llvm/TargetParser/Host.h"
+
+#include <fstream>
+
+// a lot of this code is adapted from IREE's LLVMCPU target code
+namespace mlir::tt::llvm_to_cpu {
+
+// helper to create randomized tempDir to store our temp files
+llvm::SmallString<128> createTempDir() {
+  llvm::SmallString<128> tempDir;
+  if (llvm::sys::fs::createUniqueDirectory("ttmlir_tmp", tempDir)) {
+    llvm::errs() << "Error: Could not create temporary directory.\n";
+    exit(1);
+  }
+  return tempDir;
+}
+
+// helper to create specific temp file inside a dir
+llvm::SmallString<128> createTempFile(llvm::StringRef tempDir,
+                                      llvm::StringRef prefix,
+                                      llvm::StringRef extension) {
+  llvm::SmallString<128> tempFile;
+  llvm::sys::path::append(tempFile, tempDir, prefix + "-%%%%%%" + extension);
```

Comment:
Is this string of `%%%%%%` conventional?

Reply:
the idea is to get a randomized path, llvm Path.cpp has code like
```
// Replace '%' with random chars.
  for (unsigned i = 0, e = ModelStorage.[size](https://llvm.org/doxygen/classllvm_1_1SmallVectorBase.html#adfb79daedf5d5b469756b95066752b38)(); i != e; ++i) {
    if (ModelStorage[i] == '%')
      ResultPath[i] = "0123456789abcdef"[[sys::Process::GetRandomNumber](https://llvm.org/doxygen/classllvm_1_1sys_1_1Process.html#a92b058c34d04f6d0adf6eac3060343c6)() & 15];
  }
  ```

Reply:
so I think this would standard way to generate randomized path, but don't have any specific example to point to